### PR TITLE
cmd/bosun: (wip) link previous incidents and next incident

### PR DESCRIPTION
**Pull request by kylebrandt**
requires a migration as all of the IncidentState objects are updated.
This makes it so when viewing incidents you can quickly navigate to previous incidents
or the next incident if there is a new one for the same alert key.

### CODE REVIEW

**Comment by kylebrandt** [comment]
selection_073

**Comment by kylebrandt** [comment]
TODO:

- Document new Template Vars available
- Add previous incident links to UI in dashboard
- Idea: Add template func to fetch other Incident states so you can do things like include the last close, notes, etc messages from the previous incident. 
 - Make it so the rule page will fetch actual incident history if your testing number matches a real incident.

**Comment by kylebrandt** [comment]
example for the template func:

```
        <h3>Previous Ack/Close Reasons</h3>
        
        <table>
        <tr>
            <th>Action</th>
            <th>Incident</th>
            <th>Who</th>
            <th>When</th>
            <th>Why</th>
        {{- range $id := .PreviousIds -}}
            {{- $pi := $.GetIncidentState $id -}}
            {{- if notNil $pi -}}
                {{- range $action := $pi.Actions -}}
                    {{- $actionString := $action.Type | printf "%s" -}}
                    {{- $closed := eq $actionString "Closed" -}}
                    {{- $ackd := eq $actionString "Acknowledged" -}}
                    {{- if and (or $closed $ackd) (ne $action.User "bosun") }}
                        <tr>
                            <td>{{ $action.Type | printf "%s" }}</td>
                            <td><a target="_blank" href="https://bosun.ds.stackexchange.com/incident?id={{$pi.Id}}">#{{$pi.Id}}</a></td>
                            <td>{{ $action.User }}</td>
                            <td>{{ $action.Time.Format "2006-01-02 15:04" }}</td>
                            <td>{{ $action.Message }}</td>
                        </tr>
                    {{- end -}}
                {{- end -}}
            {{- else -}}
                <tr><td>{{ .LastError }}</td><td></td><td></td><td></td><td></td></tr>   
            {{- end -}}
        {{- end -}}
        </table>
```

**Comment by kylebrandt** [comment]
Fancier table with col/rowspan (note to self: if making into example check for nil on incident End time just in case):

```
        <style>
            td, th {
                padding-right: 10px;
                border: 1px solid black;
            }
        </style>
        <h3>Previous Ack/Close Reasons</h3>
        
        <table>
        <thead>
            <tr>
                <th colspan="3">Incident</th>

                <th colspan="4">Actions</th>
            </tr>
            <tr>
                <th>Id</th>
                <th>Duration</th>
                <th>Event Count</th>
                <th>Who</th>
                <th>Action Type</th>
                <th>When</th>
                <th>Why</th>
            </tr>
        </thead>
        {{- range $id := .PreviousIds -}}
            {{- $pi := $.GetIncidentState $id -}}
            {{- if notNil $pi -}}
                {{- $filteredActions := makeSlice -}}
                {{- $incidentDuration := $pi.End.Sub $pi.Start -}}
                {{- range $action := $pi.Actions -}}
                    {{- $actionString := $action.Type | printf "%s" -}}
                    {{- $closed := eq $actionString "Closed" -}}
                    {{- $ackd := eq $actionString "Acknowledged" -}}
                    {{- if and (or $closed $ackd) (ne $action.User "bosun") }}
                        {{- $filteredActions = append $filteredActions $action -}}
                    {{- end -}}
                {{- end -}}
                
                <tr>
                    {{ $actionLen := len $filteredActions }}
                    <td {{ if gt $actionLen 1 -}} rowspan="{{- len $filteredActions }}" {{- end -}}>
                        <a target="_blank" href="https://bosun.ds.stackexchange.com/incident?id={{$pi.Id}}">#{{$pi.Id }}</a>    
                    </td>
                    <td {{ if gt $actionLen 1 -}} rowspan="{{- len $filteredActions }}" {{- end -}}>
                        {{ $incidentDuration.Truncate 1e9 }}
                    </td>
                    <td {{ if gt $actionLen 1 -}} rowspan="{{- len $filteredActions }}" {{- end -}}>
                        {{ len $pi.Events }}
                    </td>
                    {{- range $ia, $action := $filteredActions -}}
                        {{- if gt $ia 0 -}}</tr><tr></td>{{ end }}
                        <td>{{ $action.User }}</td>
                        <td>{{ $action.Type | printf "%s" }}</td>
                        <td>{{ $action.Time.Format "2006-01-02 15:04" }}</td>
                        <td>{{ $action.Message }}</td>
                        {{ if gt $ia 0 }}</tr>{{ end }}
                    {{- end -}}
                </tr>
                
            {{- else -}}
                <tr><td rowspan=7>{{ .LastError }}</td></tr>   
            {{- end -}}
        {{- end -}}
        </table>
```

File: cmd/bosun/database/migrate.go [hunk 1]
```diff
@@ -75,6 +81,52 @@ func migrateRenderedTemplates(d *dataAccess) error {
...
+	conn := d.Get()
+	defer conn.Close()
+
+	for _, id := range ids {
+		incident, err := d.State().GetIncidentState(id)
+		if err != nil {
+			return err
+		}
+		previousIds := []int64{}
+		previousIds, err = d.State().GetAllIncidentIdsByAlertKey(incident.AlertKey)
```
Comment (captncraig) [inline cmd/bosun/database/migrate.go]: At least cache these results by alert key.

File: cmd/bosun/sched/check.go [hunk 2]
```diff
@@ -244,10 +245,34 @@ func (s *Schedule) runHistory(r *RunHistory, ak models.AlertKey, event *models.E
...
 				return
 			}
+			previousIds := []int64{}
+			previousIds, err = daState.GetAllIncidentIdsByAlertKey(ak)
+			if err != nil {
+				return
+			}
+			for _, id := range previousIds {
+				if incident.Id > id {
+					incident.PreviousIds = append(incident.PreviousIds, id)
```
Comment (captncraig) [inline cmd/bosun/sched/check.go]: Wait, we're storing **all** previous alert keys on **every** alert? Gross. Why?

Comment (kylebrandt) [reply cmd/bosun/sched/check.go]: @captncraig To make it easier to either:
 1. Display more than just the previous incident in a view
 2. Make it easier to iterate over all previous incidents in templates without having to follow a linked list.

Comment (kylebrandt) [reply cmd/bosun/sched/check.go]: Also to clarify not the previous alert keys, just an array of incident id numbers (int64)

Comment (kylebrandt) [reply cmd/bosun/sched/check.go]: Another Also :P alert keys are the alert name plus the tagset (not all alerts under the name, unless the tagset is `{}` or of len 1. So the list generally should not be crazy long.

Comment (captncraig) [reply cmd/bosun/sched/check.go]: I guess I'm just uncomfortable anytime I see potentially large, and also redundant, and also repetitive collections or lists. 

I guess though I'd rather a template could show them all without having to recursively traverse the chain, which was gonna be the alternative I suggested. So ok.

**Pull request merged**